File: 06-arquitectura-microservicios-java/microservicios/servicio-usuarios/src/main/java/com/portfolio/arquitectura/usuarios/dominio/objetoValor/CorreoElectronico.java
package com.portfolio.arquitectura.usuarios.dominio.objetoValor;

import lombok.EqualsAndHashCode;
import lombok.Getter;

import javax.validation.constraints.Email;
import javax.validation.constraints.NotBlank;
import java.util.regex.Pattern;

/**
 * Objeto Valor para Correo Electrónico
 * 
 * Implementa validación de correo electrónico y encapsula la lógica de negocio
 * relacionada con direcciones de correo electrónico.
 * 
 * @author Gabriel - Arquitecto de Software
 */
@Getter
@EqualsAndHashCode
public class CorreoElectronico {

    private static final Pattern PATRON_CORREO = Pattern.compile(
        "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"
    );

    @NotBlank
    @Email
    private final String valor;

    /**
     * Constructor privado para crear instancias válidas
     * 
     * @param correo Dirección de correo electrónico
     * @throws IllegalArgumentException si el correo no es válido
     */
    private CorreoElectronico(String correo) {
        if (correo == null || correo.trim().isEmpty()) {
            throw new IllegalArgumentException("Correo electrónico no puede ser nulo o vacío");
        }
        
        String correoLimpio = correo.trim().toLowerCase();
        
        if (!PATRON_CORREO.matcher(correoLimpio).matches()) {
            throw new IllegalArgumentException("Formato de correo electrónico inválido: " + correo);
        }
        
        this.valor = correoLimpio;
    }

    /**
     * Método de fábrica para crear un CorreoElectronico válido
     * 
     * @param correo Dirección de correo electrónico
     * @return Instancia de CorreoElectronico
     * @throws IllegalArgumentException si el correo no es válido
     */
    public static CorreoElectronico de(String correo) {
        return new CorreoElectronico(correo);
    }

    /**
     * Verifica si el correo es de un dominio específico
     * 
     * @param dominio Dominio a verificar
     * @return true si el correo es del dominio especificado
     */
    public boolean esDelDominio(String dominio) {
        return valor.endsWith("@" + dominio.toLowerCase());
    }

    /**
     * Obtiene el dominio del correo electrónico
     * 
     * @return Dominio del correo electrónico
     */
    public String obtenerDominio() {
        int indiceArroba = valor.indexOf('@');
        return indiceArroba > 0 ? valor.substring(indiceArroba + 1) : "";
    }

    /**
     * Obtiene la parte local del correo (antes del @)
     * 
     * @return Parte local del correo electrónico
     */
    public String obtenerParteLocal() {
        int indiceArroba = valor.indexOf('@');
        return indiceArroba > 0 ? valor.substring(0, indiceArroba) : valor;
    }

    /**
     * Verifica si el correo es un correo corporativo
     * 
     * @return true si es un correo corporativo
     */
    public boolean esCorreoCorporativo() {
        String[] dominiosCorporativos = {
            "gmail.com", "yahoo.com", "hotmail.com", "outlook.com"
        };
        
        for (String dominio : dominiosCorporativos) {
            if (esDelDominio(dominio)) {
                return false;
            }
        }
        
        return true;
    }

    @Override
    public String toString() {
        return valor;
    }
}

File: 11-sistema-notificaciones-go/internal/dominio/objetoValor/correo_electronico.go
package objetoValor

import (
	"regexp"
	"strings"
)

// CorreoElectronico representa un correo electrónico válido
type CorreoElectronico struct {
	valor string
}

var (
	patronCorreo = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
)

// NuevoCorreoElectronico crea una nueva instancia de CorreoElectronico
func NuevoCorreoElectronico(correo string) (*CorreoElectronico, error) {
	if correo == "" {
		return nil, NewErrorValidacion("Correo electrónico no puede estar vacío")
	}

	correoLimpio := strings.TrimSpace(strings.ToLower(correo))

	if !patronCorreo.MatchString(correoLimpio) {
		return nil, NewErrorValidacion("Formato de correo electrónico inválido")
	}

	return &CorreoElectronico{valor: correoLimpio}, nil
}

// ObtenerValor retorna el valor del correo electrónico
func (c *CorreoElectronico) ObtenerValor() string {
	return c.valor
}

// ObtenerDominio retorna el dominio del correo electrónico
func (c *CorreoElectronico) ObtenerDominio() string {
	partes := strings.Split(c.valor, "@")
	if len(partes) != 2 {
		return ""
	}
	return partes[1]
}

// ObtenerParteLocal retorna la parte local del correo electrónico
func (c *CorreoElectronico) ObtenerParteLocal() string {
	partes := strings.Split(c.valor, "@")
	if len(partes) != 2 {
		return c.valor
	}
	return partes[0]
}

// EsDelDominio verifica si el correo es de un dominio específico
func (c *CorreoElectronico) EsDelDominio(dominio string) bool {
	return strings.HasSuffix(c.valor, "@"+strings.ToLower(dominio))
}

// EsCorreoCorporativo verifica si es un correo corporativo
func (c *CorreoElectronico) EsCorreoCorporativo() bool {
	dominiosPersonales := []string{
		"gmail.com", "yahoo.com", "hotmail.com", "outlook.com",
		"live.com", "msn.com", "aol.com", "icloud.com",
	}

	for _, dominio := range dominiosPersonales {
		if c.EsDelDominio(dominio) {
			return false
		}
	}

	return true
}

// String implementa la interfaz Stringer
func (c *CorreoElectronico) String() string {
	return c.valor
}

// Equals verifica si dos correos electrónicos son iguales
func (c *CorreoElectronico) Equals(otro *CorreoElectronico) bool {
	if otro == nil {
		return false
	}
	return c.valor == otro.valor
}
